impl/pkg/storage/storage.go

```diff
@@ -1,180 +1,36 @@
 package storage
 
 import (
-	"bytes"
 	"context"
 	"fmt"
-	"time"
+	"net/url"
 
-	"github.com/pkg/errors"
-	"github.com/sirupsen/logrus"
-	bolt "go.etcd.io/bbolt"
+	"github.com/TBD54566975/did-dht-method/pkg/storage/db/bolt"
+	"github.com/TBD54566975/did-dht-method/pkg/storage/db/postgres"
+	"github.com/TBD54566975/did-dht-method/pkg/storage/pkarr"
 )
 
-type Storage struct {
-	db *bolt.DB
+type Storage interface {
+	WriteRecord(ctx context.Context, record pkarr.PkarrRecord) error
+	ReadRecord(ctx context.Context, id string) (*pkarr.PkarrRecord, error)
+	ListRecords(ctx context.Context) ([]pkarr.PkarrRecord, error)
 }
 
-func NewStorage(path string) (*Storage, error) {
-	if path == "" {
-		return nil, errors.New("path is required")
-	}
-	db, err := bolt.Open(path, 0600, &bolt.Options{Timeout: 3 * time.Second})
+func NewStorage(uri string) (Storage, error) {
+	u, err := url.Parse(uri)
 	if err != nil {
 		return nil, err
 	}
-	return &Storage{db: db}, nil
-}
-
-// URI return filepath of boltDB,
-func (s *Storage) URI() string {
-	return s.db.Path()
-}
-
-// IsOpen return if db was opened
-func (s *Storage) IsOpen() bool {
-	if s.db == nil {
-		return false
-	}
-	return s.db.Path() != ""
-}
-
-func (s *Storage) Close() error {
-	return s.db.Close()
-}
-
-func (s *Storage) Exists(_ context.Context, namespace, key string) (bool, error) {
-	exists := true
-	var result []byte
-
-	err := s.db.View(func(tx *bolt.Tx) error {
-		bucket := tx.Bucket([]byte(namespace))
-		if bucket == nil {
-			exists = false
-			return nil
-		}
-		result = bucket.Get([]byte(key))
-		return nil
-	})
-
-	if result == nil {
-		exists = false
+	switch u.Scheme {
+	case "bolt":
+		filename := u.Host
+		if u.Path != "" {
+			filename = fmt.Sprintf("%s/%s", filename, u.Path)
+		}
+		return bolt.NewStorage(filename)
+	case "postgres":
+		return postgres.NewPostgres(uri)
+	default:
+		return nil, fmt.Errorf("unsupported db type %s (from uri %s)", u.Scheme, uri)
```

Comment:
let's make bolt the default case too so we don't fail

Reply:
How about just adding bolt for the case where the scheme is `""` (which usually means a file path was provided, not a URI)? It would be pretty confusing for ie. `mysql://user:pass@localhost/diddht` to produce a boltdb file rather than an error.

Reply:
sgtm